[ODST-180] Implementation - CSV imports as a feature in next Learning Standards Sync release

CSR2017 opened the pull request:
Please help review and merge the changes.
Ticket: https://tracker.ed-fi.org/browse/ODST-180
Thanks

CSR2017 committed:
Add sync from csv model classes

CSR2017 committed:
Add meta data retriever, mapping and synchronization interfaces

CSR2017 committed:
Define data mapping, csv data retrieval and data synchronization services

CSR2017 committed:
Add ILearningStanardCLIApplicationBase for instantiate with correct application class conditionally.

CSR2017 committed:
Add Sync from csv specific options and mapping.json

CSR2017 committed:
Add csvhelper, csharp package references

CSR2017 committed:
Add mapping.json

CSR2017 committed:
Add service collection extension for registering sync from csv specific services. update plugin connector to resolve the csv specific services.

CSR2017 committed:
Add test data file

CSR2017 committed:
Add csv sync application tests, file processor tests and config validator tests

CSR2017 committed:
Refactor csv sync application validation error handling

CSR2017 committed:
Update test project with test-data.csv file

CSR2017 committed:
Move Type conversion exception as separate class

CSR2017 committed:
Add unit tests for Data mapping process

CSR2017 committed:
Adding swagger meta data uri helper to form meta data url.

CSR2017 committed:
Refactor the LS resource retrieval from swagger document.

CSR2017 committed:
Force to support ODS / API version 3 and error while user passing version 2.

CSR2017 committed:
Refactor swagger document and metdata.json file handling

CSR2017 committed:
Add and update required unit tests and integration tests

CSR2017 committed:
Add Learning standard csv synchronizer tests.

instructurejryan commented on src/EdFi.Admin.LearningStandards.Core/Services/FromCsv/CsvLearningStandardRetriever.cs:
```diff
@@ -0,0 +1,95 @@
+            remove => _processCount -= value;
+        }
+
+        private EventHandler<AsyncEnumerableOperationStatus> _processCount;
```
Trivial, but check the formatting guide for the location on this. Might need to be up top.

instructurejryan commented on src/EdFi.Admin.LearningStandards.Core/Services/FromCsv/DataMappingProcess.cs:
```diff
@@ -0,0 +1,175 @@
+        {
+            if (string.IsNullOrEmpty(DataMappingsFilePath))
+            {
+                DataMappingsFilePath = ".//Mappings.json";
```
I would store the string value as a private field, incase this needs to become configurable in the future (through the constructor, method, etc.).

instructurejryan commented on src/EdFi.Admin.LearningStandards.Core/Models/FromCsv/LearningStandardMetaData.cs:
```diff
@@ -0,0 +1,30 @@
+    public class LearningStandardMetaData
+    {
+        public string Name { get; set; }
+        public string DataType { get; set; }
```
If these need to be strings, I would try to have a set of string constants for the known data-types so far, like "array". If storage type doesn't matter, perhaps an actual `Type`, or `enum` instead? I think there are some type conversions below; might help a bit? Just thinking out loud.

instructurejryan commented on src/EdFi.Admin.LearningStandards.Core/Services/FromCsv/DataMappingProcess.cs:
```diff
@@ -0,0 +1,175 @@
+                var nodeMetadata = nodeMetadatas.Single(m => m.Name == node.Name);
+
+
+                if (nodeMetadata.DataType == "array")
```
This is a good usage example of the comment above regarding constants or an enum.

instructurejryan commented on src/EdFi.Admin.LearningStandards.Core/Services/FromCsv/MetaDataRetriever.cs:
```diff
@@ -0,0 +1,157 @@
+        private readonly ILogger<MetaDataRetriever> _logger;
+        private readonly ISwaggerDocumentRetriever _swaggerDocumentRetriever;
+
+        private string Folder { get; set; }
```
I think this can be a field

instructurejryan commented on src/EdFi.Admin.LearningStandards.Core/Services/FromCsv/MetaDataRetriever.cs:
```diff
@@ -0,0 +1,157 @@
+
+        private string Filename => Path.Combine(Folder, "metadata.json");
+
+        private bool MetadataExists => File.Exists(Filename);
```
I think these two can stay as properties seeing they're acting as method proxies. 👍

instructurejryan commented on src/EdFi.Admin.LearningStandards.Tests/FromCsv/CsvLearningStandardRetrieverTests.cs:
```diff
@@ -0,0 +1,127 @@
+using Moq;
+using NUnit.Framework;
+
+namespace EdFi.Admin.LearningStandards.Tests.FromCsv
```
I think the formatting guide requires AAA comments on these? Maybe not

instructurejryan commented on src/EdFi.Admin.LearningStandards.Tests/FromCsv/DataMappingProcessTests.cs:
```diff
@@ -0,0 +1,408 @@
+using EdFi.Admin.LearningStandards.Core.Services.FromCsv;
+using NUnit.Framework;
+
+namespace EdFi.Admin.LearningStandards.Tests.FromCsv
```
These will probably need AAA comments too

instructurejryan commented on src/EdFi.Admin.LearningStandards.Tests/FromCsv/LearningStandardsCsvSynchronizerTests.cs:
```diff
@@ -0,0 +1,140 @@
+using Newtonsoft.Json.Linq;
+using NUnit.Framework;
+
+namespace EdFi.Admin.LearningStandards.Tests.FromCsv
```
Same as above on AAA comments

instructurejryan commented on src/EdFi.Admin.LearningStandards.CLI/Internal/LearningStandardsCLIApplication.cs:
```diff
@@ -43,11 +41,11 @@ internal LearningStandardsCLIApplication(Action<IServiceCollection> serviceSetup
+        public bool Unattended => _options.Unattended;
 
-        internal bool Unattended => _options.Unattended;
+        public LearningStandardsCLIWriter CliWriter { get; set; }
```
Are these used outside the class, or could they be private/internal?

CSR2017 replied:
yes, those properties are used in program.cs file

instructurejryan reviewed (approved):
Everything looks good! Just a few small comments - up to you whether you'd like to use them or not.

CSR2017 committed:
Fix team city build errors. Revert latest C# version features

CSR2017 committed:
Added DataTypes class to avoid type name comparisons. Applied AAA pattern on tests.

CSR2017 committed:
Remove Csharp package

CSR2017 committed:
Add validation flow for invalid rows

CSR2017 committed:
Provide different test data rows and row with parentLearningStandardReference

CSR2017 committed:
Read me files update.

temp

temp

temp

Add identificationCodes section mapping and formatting
